python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -40,14 +51,58 @@
 UPSCALING_INITIAL_NUM_NODES = 5
 
 
+class UtilizationScore(Protocol):
+    """This fancy class just defines the `UtilizationScore` protocol to be
+    some type which is well ordered (e.g. things that can be sorted).
+
+    What we're really trying to express is
+
+    ```
+    UtilizationScore = TypeVar("UtilizationScore", bound=Comparable["UtilizationScore"])
+    ```
+
+    but Comparable isn't a real type and, and a bound with a type argument
+    can't be enforced (f-bounded polymorphism with contravariance). See Guido's
+    comment for more details: https://github.com/python/typing/issues/59.
+
+    """
+
+    @abstractmethod
+    def __eq__(self, other: "UtilizationScore") -> bool:
+        pass
+
+    @abstractmethod
+    def __lt__(self: "UtilizationScore", other: "UtilizationScore") -> bool:
+        pass
+
+    def __gt__(self: "UtilizationScore", other: "UtilizationScore") -> bool:
+        return (not self < other) and self != other
+
+    def __le__(self: "UtilizationScore", other: "UtilizationScore") -> bool:
+        return self < other or self == other
+
+    def __ge__(self: "UtilizationScore", other: "UtilizationScore") -> bool:
+        return not self < other
+
+
+class UtilizationScorer(Protocol):
+    def __call__(
```

Comment:
I think it's a problem already present in the code, but it's a little hard to understand the meanings of `node_resources` and `resources`. (I understand what's going on, but it should be easier for someone just looking at function signatures and doc strings.)

Reply:
tried renaming and making the variable names a tad more obvious, but only a tad.